src/twisted/trial/_asynctest.py

```diff
@@ -190,22 +190,24 @@ def _ebDeferTearDown(self, failure, result):
             result.stop()
         self._passed = False
 
+    @defer.inlineCallbacks
     def deferRunCleanups(self, ignored, result):
         """
         Run any scheduled cleanups and report errors (if any to the result
         object.
         """
-        d = self._runCleanups()
-        d.addCallback(self._cbDeferRunCleanups, result)
-        return d
+        failures = []
+        for func, args, kwargs in self._cleanups[::-1]:
+            try:
+                yield func(*args, **kwargs)
+            except Exception:
+                failures.append(failure.Failure())
 
-    def _cbDeferRunCleanups(self, cleanupResults, result):
-        for flag, testFailure in cleanupResults:
-            if flag == defer.FAILURE:
-                result.addError(self, testFailure)
-                if testFailure.check(KeyboardInterrupt):
-                    result.stop()
-                self._passed = False
+        for f in failures:
+            result.addError(self, f)
+            if f.check(KeyboardInterrupt):
+                result.stop()
```

Comment:
```suggestion
```

I think this is unreachable because KeyboardInterrupt won't be appended to failures

Reply:
Indeed as you remarked on IRC, `KeyboardInterrupt` does not inherit from `Exception`.